### DISCUSSION TITLE
Migrating from v1 to v2

### DISCUSSION THREAD

**Discussion by internettrans** [single-spa/create-single-spa#248] [General]
# Steps to upgrade

[create-single-spa@2.0 release](https://github.com/single-spa/create-single-spa/releases/tag/v2.0.0)
[Webpack 5 announcement post](https://webpack.js.org/blog/2020-10-10-webpack-5-release/)

1. Upgrade to webpack 5, webpack-dev-server 4, and webpack-cli 4

```sh
npm install --save-dev webpack@latest webpack-dev-server@next webpack-cli@latest
```

2. Find any dependencies in your package.json that start with `webpack-config-single-spa` or `ts-config-single-spa`. Upgrade them to the new version.

```sh
# DO NOT run all of these in your project. Only run the commands that apply to your project. 

npm install --save-dev webpack-config-single-spa@latest
npm install --save-dev webpack-config-single-spa-ts@latest
npm install --save-dev webpack-config-single-spa-react@latest
npm install --save-dev webpack-config-single-spa-react-ts@latest
npm install --save-dev ts-config-single-spa@latest
```

3. Upgrade all [webpack loaders](https://webpack.js.org/loaders/#root) and [webpack plugins](https://webpack.js.org/concepts/plugins/#root) that are dependencies in your package.json. Loaders are dependencies whose names often end with `-loader`. Plugins are dependencies whose names often include `webpack-plugin`.

4. Update your package.json `scripts`. All instances of `webpack-dev-server` should be replaced with `webpack serve`. All instances of `--env.thing` should be replaced with `--env thing`. You can remove the `=true` from any `--env` parts of webpack commands, as webpack 5 defaults it to true without the `=true`.

```diff
{
  "scripts": {
    "lint": "eslint src",
-   "start": "webpack-dev-server",
+   "start": "webpack serve",
    "test": "jest --passWithNoTests",
    "build": "webpack --mode=production",
-    "analyze": "webpack --mode=production --env.analyze=true",
+    "analyze": "webpack --mode=production --env analyze",
    "format": "prettier --write .",
    "watch-tests": "jest --watch",
    "coverage": "jest --coverage --passWithNoTests"
  }
}
```

5. Remove `src/set-public-path.js` from your repository, if you haven't already. Webpack public path is now set via webpack plugin, so the set-public-path.js file is unnecessary.

6. Remove `systemjs-webpack-interop`, `css-loader`, and `style-loader`, from your project, as they are now included in webpack-config-single-spa as a dependency.

```sh
npm uninstall systemjs-webpack-interop css-loader style-loader
```

7. Upgrade index.html scripts

In your root config, upgrade systemjs, single-spa, and import-map-overrides. Delete the `named-exports.min.js` file, as it's not needed in newer versions of systemjs.

```html
<!-- Old -->
<script src="https://cdn.jsdelivr.net/npm/systemjs@6.5.0/dist/system.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/systemjs@6.5.0/dist/extras/named-exports.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/systemjs@6.5.0/dist/extras/amd.min.js"></script>

<!-- New -->
<script src="https://cdn.jsdelivr.net/npm/systemjs@6.8.3/dist/system.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/systemjs@6.8.3/dist/extras/amd.min.js"></script>
```

8. Test it out!

```sh
npm start
npm run build
npm run analyze
```

# Optional steps

## Upgrade webpack-merge

```sh
npm install --save-dev webpack-merge@latest
```

Then in your webpack config, change how you import webpack-merge:

```js
// Old
const webpackMerge = require('webpack-merge');

// New
const { merge } = require('webpack-merge');
```

Then change how you call the webpack merge function:

```js
// Old
webpackMerge.smart(...)

// New
merge(...)
```

If you're overriding webpack `module.rules`, consider using [`mergeWithRules`](https://github.com/survivejs/webpack-merge#mergewithrules) to avoid duplicate rules. [Example](https://github.com/react-microfrontends/styleguide/blob/504c8516e30274fc0e3221a719d5355b14af9500/webpack.config.js#L11)

See https://github.com/survivejs/webpack-merge and https://github.com/survivejs/webpack-merge/blob/develop/CHANGELOG.md#503--2020-07-06 for more info.

## Upgrading root config

If you created your root config a long time ago, a lot has changed in both the create-single-spa v1 and v2 releases since then. Root configs now use `webpack-config-single-spa`. 

```sh
npm install --save-dev webpack-config-single-spa@latest
```

Now update your webpack.config.js, to match [this template](https://github.com/single-spa/create-single-spa/blob/main/packages/generator-single-spa/src/root-config/templates/webpack.config.js). Here's [an example](https://github.com/polyglot-microfrontends/root-config/commit/2d156fff069e55d0d18c324dbfbef07709429a02).

## style-loader and css-loader issues

If you run into issues with style-loader/css-loader, it may be caused by duplicate copies of the libraries - one in webpack-config-single-spa and one in your project. To solve, tell webpack to use the version in webpack-config-single-spa, by using `require.resolve()`

```sh
npm uninstall style-loader css-loader
```

```js
// webpack.config.js
return merge(singleSpaDefaults, {
  module: {
    rules: [
    module: {
      rules: [
        {
          test: /\.css$/i,
          use: [
            require.resolve("style-loader", {
              paths: [require.resolve("webpack-config-single-spa")],
            }),
            require.resolve("css-loader", {
              paths: [require.resolve("webpack-config-single-spa")],
            }),
          ],
        },
      ],
    ]
  }
})
```

# Stuck?

Comment here, or in the #general channel in our [slack workspace](https://join.slack.com/t/single-spa/shared_invite/zt-j4c38g92-JmV7O~35dkN1CnaBkK7NTw) to ask questions.

**Comment by internettrans** [comment]
Thread with example upgrade PRs. Feel free to reply with your own:

- https://github.com/react-microfrontends/navbar/pull/19 https://github.com/react-microfrontends/navbar/commit/f72482831cd09d9b99717dabfb218cdc6f5a847d https://github.com/react-microfrontends/navbar/commit/c1a777c770ee187cebedd0724653c771495f2af9
- https://github.com/react-microfrontends/api/pull/10 https://github.com/react-microfrontends/api/commit/8886ee79b441ae33074446f91eb81bd84e7fa636 https://github.com/react-microfrontends/api/commit/849cde43d1bf1a072c1b71b4de504fc7120d4629
- https://github.com/react-microfrontends/people/pull/19 https://github.com/react-microfrontends/people/commit/3aa764466ca5e688e96b1828114ec4186598d7db https://github.com/react-microfrontends/people/commit/4d0af0a673764fa9b311f9f98163c88be9af9426
- https://github.com/react-microfrontends/root-config/pull/21 https://github.com/react-microfrontends/root-config/commit/e790826944317e7aa3819cc04113dd71ee013213 https://github.com/react-microfrontends/root-config/commit/be5c43152d649f80f06bb3786b379aea0a43a7fa
- https://github.com/react-microfrontends/planets/pull/17
- https://github.com/react-microfrontends/styleguide/commit/504c8516e30274fc0e3221a719d5355b14af9500
- https://github.com/polyglot-microfrontends/root-config/commit/2d156fff069e55d0d18c324dbfbef07709429a02 https://github.com/polyglot-microfrontends/root-config/commit/e2194a4b2534948f974c8b001ee04f5e07ffd083
- https://github.com/polyglot-microfrontends/navbar/commit/084cea2919d43c9b3ad245da07b118b154d0c1ab https://github.com/polyglot-microfrontends/navbar/commit/c51a2e455fbec2d96b6af3b120a922b09f8e74fe
- https://github.com/polyglot-microfrontends/styleguide/commit/46fbc262f3b1667260bd26c463752664056c3e7c https://github.com/polyglot-microfrontends/styleguide/commit/c2fe2f516e2027be365ccd920e2e85a92b457ea1 https://github.com/polyglot-microfrontends/styleguide/commit/735a89a8981b97cc435e3fb0b601379dd3f252f9

**Comment by filoxo** [comment]
Additional example webpack@5 upgrade PRs:

- https://github.com/filoxo/single-spa-example-rxjs-shared-state/pull/4
- https://github.com/filoxo/single-spa-example-shared-styled-components/pull/7
- https://github.com/filoxo/single-spa-example-webpack-lazystyletag/pull/3
- https://github.com/filoxo/single-spa-html-with-js-example/pull/6

**Comment by tam81** [comment]
Hi @joeldenning ,

I wrote a shared library and used it in another app. Both are following the webpack single spa config. I am facing the issue - `ReferenceError: System is not defined` when running unit test in the app which is referring to shared code.

I stuck me several days already so any help or hint so that I can overcome this issue. 

Sorry if this is not the right channel to ask this question.

Thanks and regards,

**Reply by internettrans** [reply]
Hi @tam81, this is a good place to ask this question. Does your code call `System.import()` in it? Or is the error happening even without a `System.import()` in your code?

If your code has a System.import(), you likely will need to mock SystemJS in your project.

```js
// jest.config.js
module.exports = {
  "setupFiles": "<rootDir>/src/setup-tests.js"
}
```

```js
// setup-tests.js

global.System = {
  "import": jest.fn()
}
```

**Reply by internettrans** [reply]
Another thing @tam81, someone else reported this same issue at https://single-spa.slack.com/archives/CLB2VV0UF/p1612775178011200. You can join our single-spa slack workspace at https://join.slack.com/t/single-spa/shared_invite/zt-l2iljnpv-pW_o92mMpMR8RWfIOI6pTQ which will let you see the conversation there (in case it helps in your situation, too)

**Reply by tam81** [reply]
Hi @joeldenning ,

Thank you for your quick response. My code does not call `System.import()` but I were able to solve the problem follow your suggestion. 

One more thing, I also need to mock `System.register()` too. 
`// setup-tests.js

global.System = {
  "import": jest.fn(),
   "register": jest.fn(),
}`

Thanks

**Reply by internettrans** [reply]
Glad to hear you got it working!

**Comment by KshitijRaj09** [comment]
@joeldenning @tam81 I am also facing same problem. I have created a utility module using single-spa and using the utility-module inside my single-spa application. While running the test using jest. I am getting error `ReferenceError: System is not defined`.
utility-module present in application's node_modules as an dependency.
If you could help.
Thanks.

**Reply by EnricoMariaFabbri** [reply]
I don't know if this is your case, but for me worked telling babel to interpretate System.

in the file jest.config.js i wrote this code:

```javascript
module.exports = {
  preset: "ts-jest",
  testEnvironment: "jsdom",
  transform: {
    "^.+\\.(js|jsx|ts|tsx)$": [
      "babel-jest",
      { configFile: "./babel.config.json" },
    ],
  },
  transformIgnorePatterns: ["/node_modules/(?!(@<company-name>)/)"],
  moduleNameMapper: {
    "^@<company-name>/<module-name>$":
      "<rootDir>/node_modules/@<company-name>/<module-name>/src/<module-name>.ts",
  },
  moduleFileExtensions: ["ts", "tsx", "js", "json"],
  testMatch: ["**/?(*.)+(spec|test).[jt]s?(x)"],
  moduleDirectories: ["node_modules", "src"],
};
```
